File: archivelib-rs-tdd/src/compress/fn225.rs
use crate::compress::{CompressU16ArrayAlias, RCompressData};
use crate::support::ArrayAlias;
use std::io::{Read, Write};

impl<R: Read, W: Write> RCompressData<R, W> {
  pub fn fn225(
    &mut self,
    run_start226: i32,
    var187: &CompressU16ArrayAlias, // Can be removed?
    // self.dat_arr177: &mut [i16], // Can be removed?  &mut self.dat_arr177,
    var227: i16,
  ) {
    let mut run_start: usize = run_start226 as usize;
    let var289 = self.dat_arr177[run_start];
    loop {
      let mut run_length276 = 2 * run_start;
      if !(run_length276 <= var227 as usize) {
        break;
      }
      if run_length276 < var227 as usize
        && var187.get(self, self.dat_arr177[run_length276] as usize)
          > var187.get(self, self.dat_arr177[run_length276 + 1] as usize)
      {
        run_length276 += 1
      }
      if var187.get(self, var289 as usize)
        <= var187.get(self, self.dat_arr177[run_length276] as usize)
      {
        break;
      }
      self.dat_arr177[run_start] = self.dat_arr177[run_length276];
      run_start = run_length276
    }
    self.dat_arr177[run_start] = var289 as i16;
  }
}

File: archivelib-sys-refactored/c-lib/src/r_compress/fn225.c

#include "support/compress.h"

#include "r_compress.hpp"

void fn225(RCompressData *data, int32_t run_start226, uint16_t *_187,
           int16_t *_177, int16_t _227) {
  int32_t run_length276, _289;
  _289 = _177[run_start226];
  while ((run_length276 = 2 * run_start226) <= _227) {
    if (run_length276 < _227 &&
        _187[_177[run_length276]] > _187[_177[run_length276 + 1]]) {
      run_length276++;
    }
    if (_187[_289] <= _187[_177[run_length276]]) {
      break;
    }
    _177[run_start226] = _177[run_length276];
    run_start226 = run_length276;
  }
  _177[run_start226] = (uint16_t)_289;
}
